pkg/upstream/file.go

```diff
@@ -2,32 +2,66 @@ package upstream
 
 import (
 	"net/http"
+	"net/url"
 	"runtime"
 	"strings"
 
+	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/options"
+
 	"github.com/oauth2-proxy/oauth2-proxy/v7/pkg/apis/middleware"
 )
 
 const fileScheme = "file"
 
 // newFileServer creates a new fileServer that can serve requests
 // to a file system location.
-func newFileServer(id, path, fileSystemPath string) http.Handler {
+func newFileServer(upstream options.Upstream, fileSystemPath string) http.Handler {
+	handler := newFileServerForPath(fileSystemPath)
+
+	if upstream.RewriteTarget == "" {
+		// if the upstream does not have a rewrite target, strip off the Path prefix
+		// (so e.g. a request for /static/some-file.html looks for some-file.html
+		// relative to the fileSystemPath rather than static/some-file.html).
+		handler = http.StripPrefix(upstream.Path, handler)
+	} else {
+		// if the upstream *does* have a rewrite target then that means the target
+		// path relative to the fileSystemPath will be the one in the (rewritten)
+		// RequestURI.
+		handler = requestURIToURL(handler)
+	}
 	return &fileServer{
-		upstream: id,
-		handler:  newFileServerForPath(path, fileSystemPath),
+		upstream: upstream.ID,
+		handler:  handler,
 	}
 }
 
 // newFileServerForPath creates a http.Handler to serve files from the filesystem
-func newFileServerForPath(path string, filesystemPath string) http.Handler {
+func newFileServerForPath(filesystemPath string) http.Handler {
 	// Windows fileSSystemPath will be be prefixed with `/`, eg`/C:/...,
 	// if they were parsed by url.Parse`
 	if runtime.GOOS == "windows" {
 		filesystemPath = strings.TrimPrefix(filesystemPath, "/")
 	}
 
-	return http.StripPrefix(path, http.FileServer(http.Dir(filesystemPath)))
+	return http.FileServer(http.Dir(filesystemPath))
+}
+
+// requestURIToURL returns a Handler that replaces the URL in its request with
+// the result of parsing req.RequestURI.  This is necessary for file handlers
+// that have a rewrite target, since http.FileServer uses req.URL.Path when
+// looking for the target file, but the rewrite handler only updates the
+// RequestURI, leaving the original path in the URL.
+func requestURIToURL(handler http.Handler) http.Handler {
+	return http.HandlerFunc(func(rw http.ResponseWriter, req *http.Request) {
+		reqURL, err := url.ParseRequestURI(req.RequestURI)
+		if err != nil {
+			http.Error(rw, "500 Internal Server Error", http.StatusInternalServerError)
+			return
+		}
+
+		req.URL = reqURL
+		handler.ServeHTTP(rw, req)
+	})
```

Comment:
Question - is it better to fix this up here just for `file:` upstreams or would it make more sense to fix the rewrite middleware to adjust `req.URL` as well as `req.RequestURI`?

Reply:
Can you see if we use `req.URL` elsewhere? I'm wondering if past us left that alone so we had the original available _after_ the rewrite had happened

Reply:
It looks like the main places that use `http.Request.URL` outside of tests are:

- `extractAllowedEntities` in `oauthproxy.go` uses `req.URL.Query()` but not the path
- `OAuthProxy.OAuthStart` uses the `Query()` from the `/oauth2/start` URL as a source for additional parameters to pass on to the provider's login URL (this was a previous PR of mine to allow for runtime overriding of things like `?prompt=consent`)
- `app/redirect/getters.go` uses `req.URL.RequestURI()` to determine where to redirect back to when a user gets auto-redirected to the login page when hitting a protected URL without a session - I assume this happens at an earlier stage in the handler chain, as the rewrite middleware only kicks in once we know they're allowed to see a page and we want to work out where to proxy to on the upstream
- the  CSRF cookie uses `req.URL.Query()` to extract the `state` parameter, which I think only applies during the OAuth flow, not on requests that will involve an upstream
- health check and readiness check middleware checks `req.URL.EscapedPath()` to see whether an incoming request is a health probe
- `redirectToHTTPS` builds a new URL from `req.URL.String()` and substitutes the scheme before redirecting the caller
- the request logger checks `url.Path` against a list of excluded paths - I guess this needs the original pre-rewrite path so where in the chain does the logger sit, is it before or after the rewrite middleware?
- `pkg/upstream/http.go` `setProxyDirector` copies `req.RequestURI` over to `req.URL.Opaque`

and the rewrite middleware itself of course.

Reply:
@JoelSpeed I guess if we're in any way unsure then best to leave it the way I've done it here, only fixing up the `req.URL` specifically when it's about to go into the `http.FileServer` handler.